File: netDxf/IO/DxfWriter.DataTable.cs
using System.Linq;
using netDxf.Collections;
using netDxf.Objects;

namespace netDxf.IO
{
    internal sealed partial class DxfWriter
    {
        private void PrepareDataTableClass(DxfClassCollection definitions)
        {
            this.PrepareStoredEnvelopeClass(definitions, "DATATABLE", "AcDbDataTable", "ObjectDBX Classes", 0,
                this.doc.Objects.Items.Any(item => item is DxfDataTable));
        }
        private bool WriteDataTablePayload(DxfDatabaseObject item)
        {
            if (!(item is DxfDataTable table)) return false;
            this.chunk.Write(100, "AcDbDataTable");
            this.chunk.Write(70, table.StoredVersion);
            this.chunk.Write(90, table.Columns.Count);
            this.chunk.Write(91, table.RowCount);
            this.chunk.Write(1, this.EncodeDatabaseString(table.Name));
            foreach (DxfDataColumn column in table.Columns)
            {
                this.chunk.Write(92, (int)column.Type);
                this.chunk.Write(2, this.EncodeDatabaseString(column.Name));
                foreach (object value in column.Values)
                {
                    switch (column.Type)
                    {
                        case DxfDataCellType.Integer: this.chunk.Write(93, (int)value); break;
                        case DxfDataCellType.Double: this.chunk.Write(40, (double)value); break;
                        case DxfDataCellType.String: this.chunk.Write(3, this.EncodeDatabaseString((string)value)); break;
                        case DxfDataCellType.Boolean: this.chunk.Write(71, (short)((bool)value ? 1 : 0)); break;
                        case DxfDataCellType.Point: case DxfDataCellType.Vector:
                            Vector3 point = (Vector3)value; short first = column.Type == DxfDataCellType.Point ? (short)10 : (short)11;
                            this.chunk.Write(first, point.X); this.chunk.Write((short)(first + 10), point.Y); this.chunk.Write((short)(first + 20), point.Z); break;
                        default:
                            short code = column.Type == DxfDataCellType.ObjectId ? (short)331 : column.Type == DxfDataCellType.HardOwner ? (short)360 : column.Type == DxfDataCellType.SoftOwner ? (short)350 : column.Type == DxfDataCellType.HardPointer ? (short)340 : (short)330;
                            this.chunk.Write(code, ((DxfObject)value)?.Handle ?? "0"); break;
                    }
                }
            }
            return true;
        }
    }
}

File: javascript/netDxf/IO/DxfWriter.DataTable.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import { DxfDataTable } from '../Objects/DxfDataTable.js';import { PrepareStoredEnvelopeClass } from './DxfWriter.LayerFilterPointer.js';
import { EncodeDxfDatabaseText } from '../../runtime/DxfStringEncoding.js';
export function PrepareDataTableClass(document,definitions){
  PrepareStoredEnvelopeClass(document,definitions,'DATATABLE','AcDbDataTable','ObjectDBX Classes',0,Array.from(document.Objects.Items).some(item=>item instanceof DxfDataTable));
}
export function WriteDataTablePayload(chunk,version,item){
  if(!(item instanceof DxfDataTable))return false;chunk.Write(100,'AcDbDataTable');chunk.Write(70,item.StoredVersion);chunk.Write(90,item.Columns.Count);chunk.Write(91,item.RowCount);chunk.Write(1,EncodeDxfDatabaseText(item.Name,version));
  for(const column of item.Columns){
    chunk.Write(92,column.Type);chunk.Write(2,EncodeDxfDatabaseText(column.Name,version));
    for(const value of column.Values){
      switch(column.Type){
        case 1:chunk.Write(93,value);break;case 2:chunk.Write(40,value);break;case 3:chunk.Write(3,EncodeDxfDatabaseText(value,version));break;case 10:chunk.Write(71,value?1:0);break;
        case 4:case 11:{const first=column.Type===4?10:11;chunk.Write(first,value.X);chunk.Write(first+10,value.Y);chunk.Write(first+20,value.Z);break;}
        default:chunk.Write(column.Type===5?331:column.Type===6?360:column.Type===7?350:column.Type===8?340:330,value?.Handle??'0');break;
      }
    }
  }
  return true;
}
